x/delegation/keeper/delegation_state.go

```
@@ -141,3 +145,58 @@ func (k Keeper) GetDelegationInfo(ctx sdk.Context, stakerID, assetID string) (*d
 
 	return &ret, nil
 }
+
+// DelegationStateByOperatorAssets get the specified assets state delegated to the specified operator
+func (k *Keeper) DelegationStateByOperatorAssets(ctx sdk.Context, operatorAddr string, assetsFilter map[string]interface{}) (map[string]map[string]delegationtype.DelegationAmounts, error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), delegationtype.KeyPrefixRestakerDelegationInfo)
+	iterator := sdk.KVStorePrefixIterator(store, nil)
+	defer iterator.Close()
+
+	ret := make(map[string]map[string]delegationtype.DelegationAmounts, 0)
+	for ; iterator.Valid(); iterator.Next() {
+		var amounts delegationtype.DelegationAmounts
+		k.cdc.MustUnmarshal(iterator.Value(), &amounts)
+		keys, err := stakingtypes.ParseJoinedKey(iterator.Key())
+		if err != nil {
+			return nil, err
+		}
+		if len(keys) != 3 {
+			continue
+		}
+		restakerID, assetID, findOperatorAddr := keys[0], keys[1], keys[2]
+		if operatorAddr != findOperatorAddr {
+			continue
+		}
+		_, assetIDExist := assetsFilter[assetID]
+		_, restakerIDExist := ret[restakerID]
+		if assetIDExist {
+			if !restakerIDExist {
+				ret[restakerID] = make(map[string]delegationtype.DelegationAmounts)
+			}
+			ret[restakerID][assetID] = amounts
+		}
+	}
+	return ret, nil
+}
+
+func (k *Keeper) IterateDelegationState(ctx sdk.Context, f func(restakerID, assetID, operatorAddr string, state *delegationtype.DelegationAmounts) error) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), delegationtype.KeyPrefixRestakerDelegationInfo)
+	iterator := sdk.KVStorePrefixIterator(store, nil)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var amounts delegationtype.DelegationAmounts
+		k.cdc.MustUnmarshal(iterator.Value(), &amounts)
+		keys, err := stakingtypes.ParseJoinedKey(iterator.Key())
+		if err != nil {
+			return err
+		}
+		if len(keys) == 3 {
+			err = f(keys[0], keys[1], keys[2], &amounts)
```

Would it be better to make this function return a boolean value instead of error? I have seen boolean values being used in the Cosmos SDK and not error types in most situations. If the boolean returned is false, the iteration exits. Here's an example from the `x/staking` module.

```go
func (k Keeper) IterateDelegations(ctx sdk.Context, delAddr sdk.AccAddress,
	fn func(index int64, del types.DelegationI) (stop bool),
) {
	store := ctx.KVStore(k.storeKey)
	delegatorPrefixKey := types.GetDelegationsKey(delAddr)

	iterator := sdk.KVStorePrefixIterator(store, delegatorPrefixKey) // smallest to largest
	defer iterator.Close()

	for i := int64(0); iterator.Valid(); iterator.Next() {
		del := types.MustUnmarshalDelegation(k.cdc, iterator.Value())

		stop := fn(i, del)
		if stop {
			break
		}
		i++
	}
}
```

---

I see the a lot of interfaces of state implemented in cosmos-sdk  don't have an error as a return value. I think if `IterateDelegations` returns an error,   the `fn` might also need to return an error to handle the situation in which  the `fn` may cause an error.  But I am not sure whether we should change our interfaces according to the cosmos-sdk.
So maybe I can change it when we figure out whether the implementation of cosmos-sdk is better.